xhype/xhype/src/hv/mod.rs

```diff
@@ -537,4 +538,35 @@ mod test {
         assert_eq!(space1_id, space2_id);
         vm_destroy().unwrap();
     }
+
+    /*
+        This test is meant to show that, if we set one vcpu to a non-default
+        memory space and then destroy that vcpu, then we cannot destroy this
+        memory space any more, because the memory space will still think itself
+        occupied by some vcpu and hv_vm_space_destroy() will return HV_BUSY.
+        This is why in VCPU::drop() we have to set the memory space of any vcpu
+        to the default memory space and then destroy the vcpu.
+        tested on macOS 10.15.5
+    */
+    #[test]
+    fn hv_mem_space_destroy_test() {
+        let mut ret;
+        let mut space_id = 0;
+        let mut vcpu = 0;
+        unsafe {
+            ret = hv_vm_create(HV_VM_DEFAULT);
+            assert_eq!(ret, HV_SUCCESS);
+            ret = hv_vm_space_create(&mut space_id);
+            assert_eq!(ret, HV_SUCCESS);
+            ret = hv_vcpu_create(&mut vcpu, HV_VCPU_DEFAULT);
+            assert_eq!(ret, HV_SUCCESS);
+            ret = hv_vcpu_set_space(vcpu, space_id);
+            assert_eq!(ret, HV_SUCCESS);
+            ret = hv_vcpu_destroy(vcpu);
+            assert_eq!(ret, HV_SUCCESS);
+            ret = hv_vm_space_destroy(space_id);
+            // vcpu is destroyed, but the memory space still think itself occupied by the vcpu
+            assert_eq!(ret, HV_BUSY);
```

Comment:
I am a bit confused, is this how it is supposed to work?

Reply:
This test shows an example of how to call these functions in a "bad" way. By "bad" I mean the Hypervisor framework thinks it bad. But I think it is a bug of Hypervisor framework.

I first set the vcpu to be associated with the memory space, and then the vcpu is destroyed. Since the vcpu is destroyed, we should be able to destroy the memory space. But `hv_vm_space_destroy(space_id)` will __not__ return success, instead, it returns `HV_BUSY`, indicating that it is used by some vcpu.

So if we want to destroy the memory space in a "correct" way, we have to first set the vcpu's memory space to 0 (the default memory space), then destroy the vcpu. After that we are able to destroy the memory space.